src/paginationCalls/pageDescriptor.ts

```diff
@@ -103,6 +116,96 @@ export class PageDescriptor implements Descriptor {
     return stream;
   }
 
+  /**
+   * Create an async iterable which can be recursively called for data on-demand.
+   */
+  asyncIterate(
+    apiCall: GaxCall,
+    request: RequestType,
+    options: CallSettings
+  ): AsyncIterable<{} | undefined> {
+    const iterable = this.createIterator(options);
+    const funcPromise =
+      typeof apiCall === 'function' ? Promise.resolve(apiCall) : apiCall;
+    funcPromise
+      .then((func: GaxCall) => {
+        this.makeCall(request, func, options);
+      })
+      .catch(error => {
+        throw new Error(error);
+      });
+    return iterable;
+  }
+
+  createIterator(options: CallSettings): AsyncIterable<{} | undefined> {
+    const self = this;
+    const cache = this.cache;
+    const asyncIterable = {
+      [Symbol.asyncIterator]() {
+        const paramPromise: Promise<[
+          RequestType,
+          SimpleCallbackFunction
+        ]> = new Promise((resolve, reject) => {
+          self.resolveParams = resolve;
+        });
+        let nextPageRequest: RequestType | null = {};
+        let firstCall = true;
+        return {
+          async next() {
+            const ongoingCall = new call.OngoingCallPromise(options.promise);
+            const [request, func] = await paramPromise;
+            if (cache.length > 0) {
+              return Promise.resolve({done: false, value: cache.shift()});
+            }
+            if (!firstCall && !nextPageRequest) {
+              return Promise.resolve({done: true, value: undefined});
+            }
+            nextPageRequest = await self.getNextPageRequest(
+              func,
+              firstCall ? request : nextPageRequest!,
+              ongoingCall
+            );
+            firstCall = false;
+            if (cache.length === 0) {
+              nextPageRequest = null;
+              return Promise.resolve({done: true, value: undefined});
+            }
+            return Promise.resolve({done: false, value: cache.shift()});
+          },
+        };
+      },
+    };
+    return asyncIterable; // return iterable
```

Comment:
nit: please remove this comment, it does not add anything to the code :) `asyncIterable` is a great name that perfectly shows that an iterable is returned.